networkx/algorithms/shortest_paths/weighted.py

```diff
@@ -125,7 +125,18 @@ def dijkstra_path_length(G, source, target, weight='weight'):
     --------
     bidirectional_dijkstra()
     """
-    length = dict(single_source_dijkstra_path_length(G, source, weight=weight))
+    
+    if source == target:
+        return 0
+
+    if G.is_multigraph():
+        get_weight = lambda u, v, data: min(
+            eattr.get(weight, 1) for eattr in data.values())
+    else:
+        get_weight = lambda u, v, data: data.get(weight, 1)
+    
+    length =  dict(_dijkstra(G, source, get_weight, target=target))
```

Comment:
You are completely right. As a matter of fact, when looking into it, _dijkstra() returned an iter() of an already existing dict, so it is even double the work. I changed the behaviour of _dijkstra() to return the original dict (which is safe since it is a private function), and have the few functions that really need to return the iterator create it themselves.

While at it, I also removed the logic that the function sometimes returns (pred, dist), (dist, path) or iter(dist) into always returning dist. The returned pred and path dictionaries had to be created and passed as references by the caller anyway when applicable, so these functions already have access to them. This also enables using both path and predecessor dicts simultaneously, or adding another optional successor dict.